sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala

```diff
@@ -583,7 +583,9 @@ object FunctionRegistry {
           val validParametersCount = constructors
             .filter(_.getParameterTypes.forall(_ == classOf[Expression]))
```

Comment:
If we remove this line, what is the outcome?

In this PR, the logic is to add a special case. When unable to find a constructor whose parameters are all Expressions, we use the remaining constructors. [Note, headOption means we only choose the first one] Could you please check whether we should list all of them in the error messages?

Reply:
If we remove this line, it works fine. I did this in the first commit of the pr,  discussed with HyukjinKwon [here](https://github.com/apache/spark/pull/24947/files/4af96a1b90ceb54c64c798215d2f5eefe87ad896#r296583009)

Reply:
No, It doesn't work fine.
We should only recognise the constructor with expressions because those are only able to be used in SQL. Otherwise, it shows incorrect information about possible argument combination which partially reverts #21226.

Reply:
The problem is that we have function registered `in` but that's unable/not supposed to to call via `in(...)` currently. Possibly there might be more cases like this.